docs/proposals/affinity.md

```diff
@@ -0,0 +1,341 @@
+
+* **Put the pod on a node that is already running a pod that requires a license for software package P**:
+Assuming pods that require a license for software package P have a label `{key=license, value=P}`:
+`{PodSelector: "license" In "P", TopologyKey: "node"}`
+
+* **Put this pod in the same zone as other pods from its same service**:
+Assuming pods from this pod's service have some label `{key=service, value=S}`:
+`{PodSelector: "service" In "S", TopologyKey: "zone"}`
+
+This last example illustrates a small issue with this API when it is used
+with a scheduler that processes the pending queue one pod at a time, like the current
+Kubernetes scheduler. The HardAffinityExpression
+`{PodSelector: "service" In "S", TopologyKey: "zone"}`
+only "works" once one pod from service S has been scheduled. But if all pods in service
+S have this HardAffinityExpression in their PodSpec, then the HardAffinityExpression will block the first
+pod of the service from ever scheduling, since it is only allowed to run in a zone with another pod from
+the same service. And of course that means none of the pods of the service will be able
+to schedule. This problem *only* applies to hard affinity, not soft affinity or hard or soft
+anti-affinity. There are at least two ways to solve this problem
+* short-term: when a controller creates pods from a PodTemplate, for exactly *one* of those
+pods, it should omit any hard affinity rules that select the pods of that PodTemplate.
+* long-term/speculative: controllers could present the scheduler with a group of pods from
+the same PodTemplate as a single unit. This would allow two things: (1) the scheduler could ignore
+problematic hard affinity rules rather than requiring the controller to handle it,
+and (2) the scheduler could do proper [gang scheduling](https://github.com/kubernetes/kubernetes/issues/16845)
+since it could receive an entire gang simultaneously.
+
+### Anti-affinity
+
+* **Spread the pods of this service S across nodes and zones**:
+`{{PodSelector: <selector that matches S's pods>, TopologyKey: "node"}, {PodSelector: <selector that matches S's pods>, TopologyKey: "zone"}}`
+
+* **Don't co-locate pods of this service with pods from service "evilService"**:
+`{PodSelector: selector that matches evilService's pods, TopologyKey: "node"}`
+
+* **Don't co-locate pods of this service with any other pods**:
+`{PodSelector: empty, TopologyKey: "node"}`
+
+* **Don't co-locate pods of this service with any other pods except other pods of this service**:
+Assuming pods from the service have some label `{key=service, value=S}`:
+`{PodSelector: "service" NotIn "S", TopologyKey: "node"}`
+Note that this works because `"service" NotIn "S"` matches pods with no key "service"
+as well as pods with key "service" and a corresponding value that is not "S."
+
+## Algorithm
+
+An example algorithm a scheduler might use to implement affinity and anti-affinity rules is as follows.
```

Comment:
Please send a followup PR with references if there are specific ones you'd like to see.